Src/WindowsServer/WindowsServer.Shared/ConfigurationTelemetryInitializer.cs

```diff
@@ -0,0 +1,114 @@
+﻿using System;
+using System.Text.RegularExpressions;
+using Microsoft.ApplicationInsights.Channel;
+using Microsoft.ApplicationInsights.Extensibility;
+using Microsoft.ApplicationInsights.Extensibility.Implementation;
+using Microsoft.ApplicationInsights.WindowsServer.Implementation;
+
+namespace Microsoft.ApplicationInsights.WindowsServer
+{
+    /// <summary>
+    /// A telemetry initializer that will gather configuration information.
+    /// </summary>
+    public class ConfigurationTelemetryInitializer : ITelemetryInitializer
+    {
+        private string roleName;
+        private string roleInstanceName;
+        private volatile bool updateEnvVars = true;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ConfigurationTelemetryInitializer" /> class.
+        /// </summary>
+        public ConfigurationTelemetryInitializer()
+        {
+            WindowsServerEventSource.Log.TelemetryInitializerLoaded(this.GetType().FullName);
+            AppServiceEnvironmentVariableMonitor.Instance.MonitoredAppServiceEnvVarUpdatedEvent += this.UpdateEnvironmentValues;
+        }
+
+        /// <summary>
+        /// Gets or sets the <c>cloud_RoleName</c>.
+        /// </summary>
+        /// <value>The <c>cloud_RoleName</c>.</value>
+        /// <remarks>Supports environment variable expansion with <c>%VARIABLE_NAME%</c></remarks>
+        public string RoleName { get; set; }
+
+        /// <summary>
+        /// Gets or sets the <c>cloud_RoleInstance</c>.
+        /// </summary>
+        /// <value>The <c>cloud_RoleInstance</c>.</value>
+        /// <remarks>Supports environment variable expansion with <c>%VARIABLE_NAME%</c></remarks>
+        public string RoleInstance { get; set; }
+
+        /// <summary>
+        /// Initializes <see cref="ITelemetry" /> role and node context information.
+        /// </summary>
+        /// <param name="telemetry">The telemetry to initialize.</param>
+        public void Initialize(ITelemetry telemetry)
+        {
+            if (this.updateEnvVars)
+            {
+                this.roleName = this.GetRoleName();
+                this.roleInstanceName = this.GetRoleInstanceName();
+                this.updateEnvVars = false;
+            }
+
+            if (this.roleName is string && string.IsNullOrEmpty(telemetry.Context.Cloud.RoleName))
```

Comment:
Can `this.roleName is string` ever be false if it's declared as a string?

Reply:
Is this just shorthand for a (not-)null check? If so, `!= null` seems more direct (not sure if the C# compiler/JIT will _actually,_ do a type check), clear and concise.

Reply:
I this case, it's more a matter of style.

But a variable of type `string` can be `null` which is not a `string`.

A construct like `stringThatMayBeNill is string stringThatIsDefinitelyNotNull` guarantees that `stringThatIsDefinitelyNotNull` is definitely not `null`.

This does not happen with `string` but any type can redefine the `==` and `!=` operators and the `true` and `false` operators in a way that `value == null` is not guaranteed `value` is actually `null`. On the other hand, the compiler guarantees that `value is null` will be resolved as `object.ReferenceEquals(value, null)`.